### PR TITLE
Replot on drag

### PR DESCRIPTION
**Pull request by etpinard**
an attempt at fixing https://github.com/plotly/plotly.js/issues/3305

In brief, this PR completely :hocho: 

https://github.com/plotly/plotly.js/blob/7bb5daaaa118537ca01464d12cdbcef6f9cc764a/src/plot_api/plot_api.js#L139-L149

and its `replotPending` logic and replaces it with a `Plots.redrag` handler called at the end `Plotly.plot` similar to `Plots.rehover`. Bye bye uncaught promises and console errors :wave: 

The current solution most likely **will break** in some scenarios, but it is enough to fix (I think) all the problematic examples listed in https://github.com/plotly/plotly.js/issues/3305#issuecomment-478748135, see:

- https://codepen.io/etpinard/pen/VNLoNB
- https://codepen.io/etpinard/pen/jRbNOW?editors=1010
- https://codepen.io/etpinard/pen/rbOBNw?editors=1111
- https://codepen.io/etpinard/pen/rbOBwp
- https://codepen.io/etpinard/pen/qwOWOK
- https://codepen.io/etpinard/pen/MRwNoY?editors=1010

For example, dragging under `dragmode:'zoom'` in an `extendTraces` interval looks like:

Peek 2019-04-02 16-51

Notice: we're not trying to resize the zoombox, but instead we update its corresponding data-space range as new (auto-ranged) data pts come in.

---------

cc @plotly/plotly_js

### CODE REVIEW

**Comment by jonmmease** [comment]
Awesome! Will this also work for updating images?

 The reason I ask is that I'd like to be able to eventually update this datashader example (https://plot.ly/python/change-callbacks-datashader/#image-updates-on-drag-zoom) to update the image during drag, not only on drag completion.  To do this I think we would need to pick #2606 up again to have access to the relayout events during drag.

**Comment by etpinard** [comment]
> Will this also work for updating images?

it should.

**Comment by etpinard** [comment]
Alright, I added some tests (mostly inspired by past bug reports) and after more than a day hunting around for edge cases, I'm fairly confident this solution is pretty good.

I'll tag this as reviewable.

File: src/plots/ternary/ternary.js [hunk 1]
```diff
@@ -526,7 +526,7 @@ proto.initInteractions = function() {
                 dragOptions.clickFn = clickZoomPan;
                 dragOptions.doneFn = dragDone;
                 panPrep();
-                clearSelect(zoomContainer);
```
Comment (archmoj) [inline src/plots/ternary/ternary.js]: Could we rename this thing to `zoomlayer` here? https://github.com/plotly/plotly.js/blob/8fea15620773d35f58c3ee6d6db00898e75d514f/src/plots/ternary/ternary.js#L496

File: src/plots/cartesian/select.js [hunk 2]
```diff
@@ -669,7 +670,7 @@ function updateSelectedState(gd, searchTraces, eventData) {
     // before anything else, update preGUI if necessary
     for(i = 0; i < searchTraces.length; i++) {
         var fullInputTrace = searchTraces[i].cd[0].trace._fullInput;
-        var tracePreGUI = gd._fullLayout._tracePreGUI[fullInputTrace.uid];
+        var tracePreGUI = gd._fullLayout._tracePreGUI[fullInputTrace.uid] || {};
```
[the remark this answers is not in the document]
Comment (etpinard) [reply src/plots/cartesian/select.js]: Absolutely not. If `gd._fullLayout` is undefined something terrible wrong must have happen, and stuff **should** fail.

File: src/plots/cartesian/select.js [hunk 3]
```diff
@@ -774,11 +775,15 @@ function fillSelectionItem(selection, searchInfo) {
...
-function clearSelect(zoomlayer) {
-    // until we get around to persistent selections, remove the outline
-    // here. The selection itself will be removed when the plot redraws
-    // at the end.
-    zoomlayer.selectAll('.select-outline').remove();
+// until we get around to persistent selections, remove the outline
+// here. The selection itself will be removed when the plot redraws
+// at the end.
+function clearSelect(gd) {
+    var fullLayout = gd._fullLayout || {};
```
[the remark this answers is not in the document]
Comment (etpinard) [reply src/plots/cartesian/select.js]: Right, this happens in our tests sometimes, when `Plotly.purge(gd)` is called after a delay.

[the remark this answers is not in the document]
Comment (etpinard) [reply src/plots/cartesian/select.js]: Notice this function clears stuff, whereas above https://github.com/plotly/plotly.js/pull/3716#discussion_r272430522 that function wants to select stuff - which can't be done w/o `gd._fullLayout`.

File: src/plot_api/plot_api.js [hunk 4]
```diff
@@ -1402,7 +1393,7 @@ function restyle(gd, astr, val, _traces) {
         seq.push(emitAfterPlot);
     }
 
-    seq.push(Plots.rehover);
+    seq.push(Plots.rehover, Plots.redrag);
```
[the remark this answers is not in the document]
Comment (etpinard) [reply src/plot_api/plot_api.js]: I'll pass.

File: src/components/dragelement/index.js [hunk 5]
```diff
@@ -13,7 +13,6 @@ var mouseOffset = require('mouse-event-offset');
 var hasHover = require('has-hover');
 var supportsPassive = require('has-passive-events');
 
-var Registry = require('../../registry');
 var Lib = require('../../lib');
```
[the remark this answers is not in the document]
Comment (etpinard) [reply src/components/dragelement/index.js]: Ok done in -> 51b74b56d

File: src/plots/ternary/ternary.js (hunk 1 above)
Comment (etpinard) [reply src/plots/ternary/ternary.js]: Ok done in -> 51b74b56d

**Comment by archmoj** [comment]
Awesome PR. 
:dancer: :dancer: :dancer:
:dancer: :dancer:
:dancer:
